### PR TITLE
SQL: make date/datetime and interval types compatible in conditional functions

### LINKED ISSUE #46674: SQL: incompatible types when using conditional functions
```
SELECT IFNULL(date, {d '2010-04-12'}) FROM test GROUP BY 1
```

Will generate an error like `2nd argument of [IFNULL(date, {d '2010-04-12'})] must be [datetime], found value [{d '2010-04-12'}] type [date]` which, while true, there shouldn't be a strict compatibility check between `date` and `datetime` field types.

### PR DESCRIPTION
**Pull request by astefan**
Until now, ES-SQL considered as compatible types: numeric data types and text data types (text and keyword). This means that something like `SELECT IFNULL(date, {d '2010-04-12'}) FROM test` would fail because `date`, being a DATETIME, was incompatible with the CASTed DATE type literal.
This PR makes date/datetime compatible and, also, expands this compatibility to INTERVAL data types, making possible queries like `SELECT hire_date + IIF(salary > 70000, INTERVAL 2 HOURS, INTERVAL 2 DAYS) FROM test`.

There is still one edge case scenario involving INTERVALs using `IN` operator (`SELECT 100 FROM test WHERE  INTERVAL 1 YEAR IN (INTERVAL 1 YEAR, INTERVAL 12 MONTH) LIMIT 3`), that will be addressed in a separate issue.

Fixes https://github.com/elastic/elasticsearch/issues/46674.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java [hunk 1]
```diff
@@ -48,7 +47,7 @@ protected TypeResolution resolveType() {
         if (resolution == TypeResolution.TYPE_RESOLVED) {
             // interval must be Literal interval
             if (field().dataType().isDateBased()) {
-                resolution = isType(interval, DataTypes::isInterval, "(Date) HISTOGRAM", ParamOrdinal.SECOND, "interval");
+                resolution = isType(interval, d -> d.isInterval(), "(Date) HISTOGRAM", ParamOrdinal.SECOND, "interval");
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java]: Isn't it enough to just do `DataType::isInterval` ?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/type/DataTypeConversionTests.java [hunk 2]
```diff
@@ -685,6 +685,6 @@ public void testIpToString() {
     }
 
     private DataType randomInterval() {
-        return randomFrom(Stream.of(DataType.values()).filter(DataTypes::isInterval).collect(Collectors.toList()));
+        return randomFrom(Stream.of(DataType.values()).filter(d -> d.isInterval()).collect(Collectors.toList()));
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/type/DataTypeConversionTests.java]: same here

**Review by matriv — APPROVED** [review]
LGTM. Just a couple of minor comments regarding the lambda expression.

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Review by costin — APPROVED** [review]
LGTM

In the future, we might consider automatically casting / widening types inside a comparison inside the analyzer so that the type checks are done generally instead of an per-operator basis.

**Pull request merged**